autogpts/autogpt/autogpt/core/resource/schema.py

```diff
@@ -39,6 +39,9 @@ def update_usage_and_cost(self, *args, **kwargs) -> None:
 class ProviderCredentials(SystemConfiguration):
     """Struct for credentials."""
 
+    def unmasked(self) -> dict:
```

Comment:
<a href="https://sourcery.ai"><img src="https://sourcery.ai/favicon.svg" height="15" alt="Sourcery logo" /></a>  **issue (complexity):** Consider simplifying the unmasking logic by integrating it directly into the classes that require it.

The introduction of the `unmask` function and the modifications to `ProviderCredentials` significantly increase the complexity of the codebase. This complexity arises from the added cognitive load required to understand the new `unmasked` method, the standalone `unmask` function, and their implications on security and data handling. Additionally, the tight coupling with Pydantic's `BaseModel` could make future maintenance more challenging.

A more streamlined approach could be to encapsulate the unmasking logic directly within the classes that require it, thus avoiding the need for a separate utility function. This not only simplifies the code but also keeps related functionalities closer together, making it easier to understand and maintain. Here's a suggested revision for `ProviderCredentials` that incorporates the unmasking logic directly:

```python
class ProviderCredentials(SystemConfiguration):
    """Struct for credentials."""

    def unmasked(self) -> dict:
        unmasked_fields = {}
        for field_name, field_value in self.dict().items():
            if isinstance(field_value, SecretStr):
                unmasked_fields[field_name] = field_value.get_secret_value()
            else:
                unmasked_fields[field_name] = field_value
        return unmasked_fields

    class Config:
        json_encoders = {
            SecretStr: lambda v: v.get_secret_value() if v else None,
            SecretBytes: lambda v: v.get_secret_value() if v else None,
            SecretField: lambda v: v.get_secret_value() if v else None,
        }
```

This approach minimizes the introduction of new complexities and keeps the handling of sensitive information within the scope of the class that requires it, adhering to the principle of encapsulation.